Feat/break axis crmod

100pah opened the pull request:
<!-- Please fill in the following information to help us review your PR more efficiently. -->

## Brief Information

This pull request is in the type of:

- [x] bug fixing
- [x] new feature
- [ ] others


See https://github.com/apache/echarts/pull/19459#issuecomment-2748824472

Ovilia reviewed (commented):
I tried with `updateAxisBreak` and found that it seems can only change `isExpanded` but not `start/end` or add non-existing breaks. So I'm not sure why it's a better idea to replace this with a `expand` or `toggle` action. Because with the current implementation, if a developer wants to implement togging breaks, he has to keep a local variable of breaks' expanding and update it like this:

```js
const initialBreaks = [{
    start: 100,
    end: 200,
    gap: 0
}, {
    start: 300,
    end: 400,
    gap: 10
}];
const localBreaks = JSON.parse(JSON.stringify(initialBreaks));
chart.on('axisBreakUpdated', function (params) {
    // Listen to changes of breaks, caused by both
    // dispatchAction and user interaction
    for (let i = 0; i < params.breaks.length; i++) {
        localBreaks[i].isExpanded = params.breaks[i].isExpanded;
    }
});

// Button click
function handleButtonClick(id) {
    localBreaks[id].isExpanded = !localBreaks[id].isExpanded;
    chart.dispatchAction({
        type: 'updateAxisBreak',
        yAxisIndex: 'all',
        breaks: localBreaks[id]
    });
}
```

Instead, if we provide `toggleAxisBreak`, it could be as simple as this and no need to keep a `localBreaks` variable and listen to `axisBreakUpdated` event:

```js
function handleButtonClick(id) {
    chart.dispatchAction({
        type: 'toggleAxisBreak',
        yAxisIndex: 'all',
        breaks: localBreaks[id]
    });
}
```

So my question is, what benefit does it bring to use `updateAxisBreak`? Maybe I missed some point.

Ovilia commented on test/axis-break.html:
```diff
@@ -102,34 +123,58 @@
+                        throw new Error('test failed');
+                    }
+                    if (params.yAxisIndex !== 0) {
+                        throw new Error('test failed');
```
It's `'all'` when click the second button. I guess this assumpting only works when interacts with the chart, instead of clicking the button, right? Will it bring misunderstanding when running this test case in the future?

Ovilia commented on test/axis-break.html:
```diff
@@ -330,9 +380,9 @@
-                                return '11:30/13:00'
-                            }
+                            formatter: (value, index) => {
+                                return `${echarts.time.format(value, '{HH}:{mm}')}`;
```
It's nice to have the axis labels adjust position as you provided in this PR.

I think there is still requirement to using a single label for breaks with zero gap. For example, in the Intraday stock chart, we may expect the break label to be `11:30/13:00` instead of two labels, like [this real-world chart](https://gushitong.baidu.com/index/ab-000001) shows:

<img width="783" alt="image" src="https://github.com/user-attachments/assets/3a70ef24-15f9-44dc-a9d4-d9ecd61fb1eb" />

I tried to hide the start label and add the start/end time in the end label, but I don't know how to make it align to center in this case.

```js
formatter: (value, index) => {
    const date = new Date(value);
    if (date.getHours() === 11 && date.getMinutes() === 30) {
        return '';
    }
    if (date.getHours() === 13 && date.getMinutes() === 0) {
        return '11:30/13:00';
    }
    return `${echarts.time.format(value, '{HH}:{mm}')}`;
}
```

<img width="563" alt="image" src="https://github.com/user-attachments/assets/528f5758-3bfe-4f85-9cc1-fd73f5827df6" />

100pah replied:
I'm not sure `11:30/13:00` is a common style in intraday stock chart, or just it's a single case.

An approach to achieve that is:
+ introduce a new option `xxxAxis.axisLabel.adjustBreakLabels: boolean` or `xxxAxis.axisLabel.breakLabelGap: number` to indicate wether adjust break labels. 
+ Users use that option to close the adjustment and use formatter function to hide one of the ends, such as 
    ```js
    axisLabel: {
        formatter: (value, index, extra) => {
            const format = echarts.time.format;
            if (!extra || !extra.break) {
                return `${format(value, '{HH}:{mm}')}`;
            }
            // PENDING: how to name them? min/max or start/end?
            if (extra.break.type === 'min') {
                return `${format(extra.break.min, '{HH}:{mm}')}/${format(extra.break.max, '{HH}:{mm}')}`;
            }
            if (extra.break.type === 'max') {
                return '';
            }
        },
    }    
    ```

Would we use that way?

Ovilia replied:
How does `adjustBreakLabels` work? I can't infer what it does from the name. Is `breakLabelGap` the gap between start/end labels? How can I know what it should be to make the break label be in the middle?

I believe `11:30/13:00` is a common intraday label and besides, it's reasonable to provide a way to show single label like `100~1000` for other cases, especially when gap is 0. If the user can only use two labels, there is no way to imply the *gap* (`~`).

Ovilia commented:
The fisheye effect is really cool. Can you further explain the real-world usage of this effect, or is this just for testing? It seems similar to using dataZoom, only that the axis is not linear. So I'm curious about how we could use this feature.

Ovilia commented:
Should we change the zrender version in package.json to be `v6` to pass the CI, or just ignore the CI?

Overall, thank you very much for this really detailed review and changes! I could never finish this feature without your help! :D

100pah commented:
> I tried with `updateAxisBreak` and found that it seems can only change `isExpanded` but not `start/end` or add non-existing breaks. So I'm not sure why it's a better idea to replace this with a `expand` or `toggle` action. Because with the current implementation, if a developer wants to implement togging breaks, he has to keep a local variable of breaks' expanding and update it like this:
> 
> So my question is, what benefit does it bring to use `updateAxisBreak`? Maybe I missed some point.

Yes, the name `updateAxisBreak` may intuitively imply that is capable of updating everything including `start`/`end`, but it actually does not. That might lead to confusion.

Should we support that update start/end in an action? Currently from my understanding, it's not necessary, since we have used start/end as the identifier of a break. If support that updating, some new identifier like `id` seems necessary, which may mades it complicated.

According to the other echarts actions, this patterns is used in several places: "doSometing" "unDoSomething" "toggleSomething", such as: 
+ actions: `select` `unselect` `toggleselect`.
+ event: all of the actions trigger the event `selectchanged`

So would we follow that pattern? like 
+ actions: `expandAxisBreak` `collapseAxisBreak` `toggleAxisBreak`
+ event: all of the actions trigger the event `axisBreakChanged`
?

100pah commented:
> The fisheye effect is really cool. Can you further explain the real-world usage of this effect, or is this just for testing? It seems similar to using dataZoom, only that the axis is not linear. So I'm curious about how we could use this feature.

To be honest I do not know the real existing scenarios. It's obtained by coincidence and I thought it might have some potential use. And it's fine if no one ends up using it since no new code was added for this feature. Basically this feature serves the similar scenario as dataZoom serving. The only slight benefit might be the fisheye preserves the overall trend.

Ovilia commented:
> Should we support that update start/end in an action? 

I don't think this should be necessary because the developer can do this with `setOption` and I don't think this should be a commonly used operation so it's OK to not support this.

`expandAxisBreak`, `collapseAxisBreak`, and `toggleAxisBreak` work for me. Either `axisBreakChanged` or `axisBreakToggled` works for me.

> echarts branch v6 might needed to merge into this PR?

I've merged v6 into this PR yesterday so I believe it's the latest code of v6 now.

I agree with the new design of `yAxisIndex` in event params.

As the name of `start/end` or `min/max`, I used `start/end` because reminds me the idea of `startAngle/endAngle` in radial axis and `start/end` and `startValue/endValue` in dataZoom, where I think the concept is the same, that it's kind of a `window` of a wider range. As for `min/max`, I think this is more like a boundary constraint rather than a selected range. But I'm open to what you think. Btw, [highcharts](https://api.highcharts.com/highcharts/yAxis.breaks) uses `from/to` pair, which may not suitable for us since we don't have such names elsewhere.

As for `ScaleBreak/AxisBreak`, I think either is OK but if I must choose one, I would prefer `AxisBreak` because it seems some other components like Timeline and Parallel are also using `Scale` but we don't support breaking for those.

100pah commented:
This set of commits (until [18b2d65](https://github.com/apache/echarts/pull/20857/commits/18b2d656d9ac781fccbaf6899ec5f30f05141fa9) ) update the following items:

## Modify the axis break action to
```ts
chart.dispatch({
    type: 'expandAxisBreak',
    // breaks identifier:
    start: xxx,
    end: xxx,
    xAxisIndex: 'all', // or number, or number[]
});
chart.dispatch({
    type: 'collapseAxisBreak',
    // breaks identifier:
    start: xxx,
    end: xxx,
    xAxisIndex: 'all', // or number, or number[]
});
chart.dispatch({
    type: 'toggleAxisBreak',
    // breaks identifier:
    start: xxx,
    end: xxx,
    xAxisIndex: 'all', // or number, or number[]
});
```
This follows the pattern of the existing actions `select`, `unselect`, `toggleSelect`.

## Refine the action event `axisbreakchanged`
All of the actions above trigger the same action `axisbreakchanged`.
This follow the format of `selectchanged` triggered (shared) by `select`, `unselect`, `toggleSelect`.

```js
chart.on('axisbreakchanged', param => {
    // param: {
    //     type: 'axisbreakchanged';
    //     // 'expandAxisBreak' | 'collapseAxisBreak' | 'toggleAxisBreak';
    //     fromAction: 'expandAxisBreak' 
    //     fromActionPayload: ..., // payload
    //     // Only include specified breaks in action.
    //     breaks: [{
    //         start: ...,
    //         end: ...
    //         xAxisIndex: 1
    //         isExpanded: true
    //         old: { isExpanded: false }
    //     }, {
    //         start: ...,
    //         end: ...
    //         xAxisIndex: 1
    //         isExpanded: true
    //         old: { isExpanded: false }
    //         ...
    //     }]
    // }    
});

// And those TS types of actions and events are exported in `export/api/core.ts`.
export {
    AxisBreakChangedEvent,
    ExpandAxisBreakPayload,
    CollapseAxisBreakPayload,
    ToggleAxisBreakPayload,
} from '../component/axis/axisAction';
```

## Introduce `refineEvent` in API `registerAction` 

Based on the existing event `selectchanged` and the new event `axisbreakchanged`, abstract the behavior to `refineEvent`. It also removes the hard code (and FIXME) of register and trigger of `selectchanged` in `echarts.ts`; and `axisbreakchanged` is implemented by it.

```ts
echarts.registerAction({
    // Already have:
    type: 'doxxx',
    event: 'xxxchanged',
    action: (payload, ecModel, api) => {
        // ...
        return someInfoForRefiningEvent;
    },
    // New added:
    refineEvent: (actionResultBatch, payload, ecModel, api) => {
        // ...
        return {eventContent};
    }
});
```

- `refineEvent` is intended to create a user-friend event that differs from the original payload, while enabling feature `connect`, and being called at the last step of the "update" procedure to ensure the complete update of all models.
- If multiple actions need to share one event name, `refineEvent` must be used. e.g., actions 'doxxx' 'undoxxx' 'togglexxx' share one event name 'xxxchanged'. And that was not able to be achieved previously. 

The rule of creating "public event" and "event for connect":
- If `refineEvent` provided, `refineEvent` creates the "public event", and "event for connect" is created internally by replicating the payload. This is because `makeActionFromEvent` requires the content of event to be the same as the original payload, while `refineEvent` creates a user-friend event that differs from the original payload.
- Else if `ActionHandler` returns an object, it is both the "public event" and the "event for connect". 
    - (@deprecated, but keep this mechanism for backward compatibility).
- Else, replicate the payload as both the "public event" and "event for connect".

This API is only for the maintainer of echarts, rather than the users of echarts.


## Fix break axis overlap with normal labels at the boundary of the axis. 
    
## Some renaming
+ ScaleBreak -> AxisBreak
+ axisLabel.formatter callback: break.type min/max -> start/end

## Introduce `axis.breakLabelLayout.moveOverlap`
`axis.breakLabelLayout.moveOverlap: boolean`
+ `true`: Default. Auto move apart the break labels if overlapping.
+ `false`: Do not handle break labels overlap. See the case `test/axis-break.html` `Introday stock`. `xAxis.breakLabelLayout.moveOverlap: false` is set and a `axisLabel.formatter` is used to hide one break label and display the entire break info like `12:00/13:00`.

## Introduce label margin configuartion
This is relevant to overlap resolving.
+ case 1: Break ends overlap. It can be moved apart a little to avoid overlap, see `breakLabelLayout.moveOverlap`.
+ case 2: Normal axis labels overlap. Existing options `axisLabel.hideOverlap`, `axisLabel.showMinValue`, `axisLabel.showMaxValue` perform the hide strategy for this case.

But text bounding rect should be able to enlarge slightly to avoid them from visually touching.
And the expansion should be able to vary in top-bottom and left-right.
- Horizontally laid out labels may appear joined if there is no enough space between them.
    - e.g., appears `200123,456` but actually `200  123,456`).
- Vertically laid-out or rotated stacked labels do not require much gap to appear neat and save space.
    - e.g. 
    ```
    200
    123,456
    ```

Thus, `axisLabel.textMargin` is introduced to provide this extra space for each label. In most case, its default value will suffice; users only need to know it when a detailed customization is required. `textMargin` follows CSS margin pattern, e.g., `axisLabel.textMargin: [2, 10]` means `[top-bottom, left-right]`. And `textMargin` is applied on the local rect of a text, rather than the rotated rect.

TL;DR: Introduce the name `textMargin` rather than reuse the existing names to avoid breaking change:
- `axisLabel.margin` historically has been used to indicate the gap between the axis and label.x/.y.
- `label.minMargin` conveys the same meaning as this `textMargin` but has a different nuance, it works like CSS margin collapse (gap = label1.minMargin/2 + label2.minMargin/2), and is applied on the rotated bounding rect rather than the original local rect.


## zrender dependency is modified to v6

## Test facilities upgrade or fixing

+ Visual test: 
    + Fix that visual test runtime thrown error may block the entire text execution without notification.
    + Recorded double mousedown without a mouseup will cause error in puppeteer. Added error message to users.
+ testHelper: 
    + Fixed some typo bugs.
